### PR TITLE
field changes to resource and resource groups

### PR DESCRIPTION
**Pull request by EvenCode**
Issue #38 and #39 are resolved here.

This commit onwards: no public method can change resource or group id.
So, any call to public functions like change_*_id () must be removed. (see resource controller)

### CODE REVIEW

File: src/resource-group.cxx [hunk 1]
```diff
@@ -13,86 +13,86 @@
...
-  return FAILED_TO_GENERATE_RANDOM_ID;
 }
 
-Resource &
-ResourceGroup::_get_resource (const int index)
+void
+ResourceGroup::create_resource_id (void)
 {
-  return *std::next(m_resources.begin (), index);
+  if (id_counter - m_resources.size() > SIZE_COUNTER_DIFFERENCE)
```
Comment (Compro-Prasad) [inline src/resource-group.cxx]: Make this a `signed` comparison.

File: src/resource-group.h [hunk 2]
```diff
@@ -34,34 +32,50 @@ class ResourceGroup
...
+
+  bool change_resource_name (const int index, const std::string & resource_name);
+  void change_resource_long_name (const int index, const std::string & resource_long_name);
   void change_resource_cost (const int index, const float & resource_cost);
+  void change_resource_description (const int index, const std::string & resource_description);
 
   const Resource & get_resource (const int index) const;
 
 private:
+  static unsigned int id_counter;
```
Comment (Compro-Prasad) [inline src/resource-group.h]: make it size_t. Why static?

Comment (EvenCode) [reply src/resource-group.h]: will do

File: src/resource-group.cxx (hunk 1 above)
Comment (r4dix10) [reply src/resource-group.cxx]: there exist no scenario when id_counter < m_resource.size(). so no need of abs(id_counter - m_resources.size())

**Comment by joesteeve** [comment]
- [ ] Rename id_counter to m_id_counter. I see that its static, but its still a member variable.
- [ ] Add a `const Resource& ResourceGroup::get (const std::string& id)`
- [ ] Add a `Resource* ResourceGroup::get_mutable (const std::string& id)`
- [ ] Remove `ResourceGroup::change_*` methods. To set these attributes, the calling code should get a `Resource` object using `ResourceGroup::get_mutable()` and modify it.
- [ ] Add a `const Resource& ResourceManager::get_resource (const std::string &id)`
- [ ] Add a `Resource* ResourceManager::get_resource_mutable (const std::string &id)`
- [ ] Add a `const ResourceGroup& ResourceManager::get_resource_group (const std::string &id)`
- [ ] Add a `Resource* ResourceManager::get_resource_group_mutable (const std::string &id)`
- [ ] Remove `ResourceManager::change*`. The calling code should get a mutable `Resource` or `ResourceGroup` and do these directly.

Please fix above and resubmit.

**Comment by EvenCode** [comment]
the `*get*` functions take `id` as argument, now `*change*` functions would need `id` in order to call the `*get*` function.

from where will `*change*` functions get `id`s ? because `*change*` is being called by controller which won't know about `id` as it is hidden from view. did i miss something here ?

`*change*` and `*get*` are taking `index` as argument as `index` of a node in linked list can be known from the view.

**Pull request merged**